Use the Windows CSPRNG to get a random seed for `Stdlib.Random`

MisterDA opened the pull request:
Windows (since Vista SP2 / Server 2008) provides a CSPRNG as the [`BCryptGenRandom`](https://learn.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom) function. Use it to generate `Stdlib.Random` default seed, mimicking the use of `getentropy` or `/dev/urandom` on Unix-like systems. See the Unix counter-part `caml_unix_random_seed`: https://github.com/ocaml/ocaml/blob/9caaced74449f2886b7e35c4befe7e0a27b77407/runtime/sys.c#L604-L644

MisterDA committed:
Use the Windows CSPRNG to get a random seed for Stdlib.Random

dustanddreams commented on runtime/sys.c:
```diff
@@ -601,7 +601,7 @@ CAMLprim value caml_sys_time(value unit)
 extern int caml_win32_random_seed (intnat data[16]);
 #else
-int caml_unix_random_seed(intnat data[16])
+int caml_unix_random_seed(intnat data[static 16])
```
This particular change seem to be orthogonal to that PR. Is it intentional, and if so, what is the rationale behind it and why isn't there a similar change in the win32 code path?

MisterDA replied:
if only… [C11: static inside array parameter square brackets](https://developercommunity.visualstudio.com/t/C11:-static-inside-array-parameter-squar/1475168)

MSVC doesn't support this C99 feature that tells that the array being passed contains at least 16 elements, which helps with type/bounds checking. I had added it to both code paths, but had to remove it on win32.

dustanddreams replied:
Can you name any C compiler/linter/checker which actually makes use of the `static` keyword in array sizes and does not only parse it? This is yet another C99 feature which looks useful on paper but hasn't found its use.

MisterDA replied:
Yes. Both gcc and clang warn if `NULL` or if an array too small is passed to the function. Interestingly gcc also warns in the non-static case if an array too small is passed. cppcheck considers in both cases that the array has a fixed size and warn of out-of-bounds accesses.

<details>

```c
void f(int a[10]);
void g(int a[static 10]);
void h(int a[static 10]) {
    a[42] = 42;
}

void m(void) {
    int a[8];
    int b[11];

    f(0);
    f(a);
    f(b);

    g(0);
    g(a);
    g(b);
}
```

```console
$ /opt/homebrew/Cellar/gcc/14.2.0_1/bin/gcc-14  -O0 -Wall -Wextra -D_FORTIFY_SOURCE=3 -c test.c
test.c: In function 'm':
test.c:15:5: warning: argument 1 null where non-null expected [-Wnonnull]
   15 |     g(0);
      |     ^
test.c:2:6: note: in a call to function 'g' declared 'nonnull'
    2 | void g(int a[static 10]);
      |      ^
test.c:12:5: warning: 'f' accessing 40 bytes in a region of size 32 [-Wstringop-overflow=]
   12 |     f(a);
      |     ^~~~
test.c:12:5: note: referencing argument 1 of type 'int[10]'
test.c:1:6: note: in a call to function 'f'
    1 | void f(int a[10]);
      |      ^
test.c:16:5: warning: 'g' accessing 40 bytes in a region of size 32 [-Wstringop-overflow=]
   16 |     g(a);
      |     ^~~~
test.c:16:5: note: referencing argument 1 of type 'int[10]'
test.c:2:6: note: in a call to function 'g'
    2 | void g(int a[static 10]);
      |      ^
$ /opt/homebrew/Cellar/llvm/19.1.2/bin/clang -O0 -Wall -Wextra -D_FORTIFY_SOURCE=3 -c test.c
test.c:15:5: warning: null passed to a callee that requires a non-null argument [-Wnonnull]
   15 |     g(0);
      |     ^ ~
test.c:2:12: note: callee declares array parameter as static here
    2 | void g(int a[static 10]);
      |            ^~~~~~~~~~~~
test.c:16:5: warning: array argument is too small; contains 8 elements, callee requires at least 10 [-Warray-bounds]
   16 |     g(a);
      |     ^ ~
test.c:2:12: note: callee declares array parameter as static here
    2 | void g(int a[static 10]);
      |            ^~~~~~~~~~~~
2 warnings generated.
```

</details>

MisterDA replied:
I've removed this bit from this PR.

nojb commented on runtime/win32.c:
```diff
@@ -622,20 +624,33 @@ void caml_win32_unregister_overflow_detection(void)
+  int nread = 0;
+
+  /* Try kernel entropy first */
+  NTSTATUS st = BCryptGenRandom(NULL, buffer, 12,
```
```suggestion
  NTSTATUS st = BCryptGenRandom(NULL, buffer, sizeof(buffer),
```

nojb commented on runtime/win32.c:
```diff
@@ -622,20 +624,33 @@ void caml_win32_unregister_overflow_detection(void)
+  NTSTATUS st = BCryptGenRandom(NULL, buffer, 12,
+                                BCRYPT_USE_SYSTEM_PREFERRED_RNG);
+  if (NT_SUCCESS(st))
+    nread = 12;
```
```suggestion
    nread = sizeof(buffer);
```

nojb commented on runtime/win32.c:
```diff
@@ -622,20 +624,33 @@ void caml_win32_unregister_overflow_detection(void)
+                                BCRYPT_USE_SYSTEM_PREFERRED_RNG);
+  if (NT_SUCCESS(st))
+    nread = 12;
+  while (nread > 0) data[n++] = buffer[--nread];
```
Why not `memcpy(data, buffer, sizeof(buffer))`?

Also, this loop can be put under the `if` above, and be made to follow by the "return" below (always unde the `if`).

MisterDA replied:
> Why not `memcpy(data, buffer, sizeof(buffer))`?

The function doesn't return the number of random bytes written to `data`, it returns the number of `intnats` that have random `unsigned char` in their low-byte. If I do a `memcpy` and then return `12`, the final OCaml seed array will have 12 intnats, of which the first 12 bytes will be random data, and the rest (I guess) the stack values that were there (or zeroes?).

nojb commented on runtime/win32.c:
```diff
@@ -622,20 +624,33 @@ void caml_win32_unregister_overflow_detection(void)
-  data[3] = pc.LowPart;
-  data[4] = pc.HighPart;
-  return 5;
+  if (n < 16) data[n++] = t.dwLowDateTime;
```
`n` is always zero if you reach here, or am I missing something?

MisterDA replied:
Yes, you're right. `BCryptGenRandom` either writes 12 bytes and succeeds; or fails and `n` will be zero.

MisterDA commented:
I agree with your comments @nojb. The idea behind the change was to copy the existing `caml_unix_random_seed` function as much as possible, which is why it looks a bit odd here (it's possible to get a partial read from `/dev/urandom` but not from `getentropy` or `BCryptGenRandom`). I guess I can simplify the code further.

https://github.com/ocaml/ocaml/blob/6e80aa02422c10c693334e29e70d425ede81e3e6/runtime/sys.c#L628-L668

One thing that surprised me with this code is that we don't fill the `intnat data[16]` buffer continuously, but sometimes only the low bytes of each `intnat`. I suppose it doesn't matter for the PRNG if the seed has zero bytes as long as some of its bytes are random.

In the end this function is a bit of a headache... so I went with the easy way, which was to copy `caml_unix_random_seed` as much as possible.

MisterDA commented:
I also wonder if `data` should be zeroed. It might get previous data from the stack. Not sure if this matters as we're still writing random(-ish) values to it.

MisterDA committed:
Light refactor of caml_unix_random_seed

- use the buffer size instead of a magic constant;
- remove unused _WIN32 guards around getpid/getppid, Windows has its
  own caml_win32_random_seed.

nojb commented on runtime/win32.c:
```diff
@@ -620,12 +622,22 @@ void caml_win32_unregister_overflow_detection(void)
+  /* Try kernel entropy first */
+  NTSTATUS st = BCryptGenRandom(NULL, buffer, sizeof(buffer),
+                                BCRYPT_USE_SYSTEM_PREFERRED_RNG);
+  if (NT_SUCCESS(st)) {
```
This is probably equivalent, but `NT_SUCCESS` checks for a range of result codes; wouldn't it be simpler to compare with `STATUS_SUCCESS` here?

MisterDA replied:
I'd rather be on the cautious side. From [Use `NTSTATUS` Values](https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/using-ntstatus-values):

> Numerous values are assigned to each type. A common mistake, when testing for a successful return from a routine, is to compare the routine's return value with `STATUS_SUCCESS`. This comparison checks for only one of several success values.

nojb replied:
OK, thanks!

nojb reviewed (commented):
As far as I can see, the patch is correct. The changes to `sys.c` could have been omitted or put in a separate PR, but am not strongly opposed to them.

I am "informally" approving this PR, as the code is tricky, so a second pair of eyes would be appreciated before merging.

LGTM

xavierleroy commented:
I remember that using bcrypt.lib was considered a long time ago, but rejected because it's a big library and linking with it would penalize all OCaml programs.  (Not sure I remember correctly.)

`Stdlib.Random` is not cryptographically strong; do we really need to link with a whole crypto library to implement it?

MisterDA commented:
> I remember that using bcrypt.lib was considered a long time ago, but rejected because it's a big library and linking with it would penalize all OCaml programs. (Not sure I remember correctly.)

I see an increase of 1-2KB on the ocaml*.exe files when building this PR.

> `Stdlib.Random` is not cryptographically strong; do we really need to link with a whole crypto library to implement it?

I think not, my intent was to have better parity with Unix which _does_ use a CSPRNG to seed `Stdlib.Random`, and the PR was also prompted by the comments suggesting better ways to gather random bytes from Windows.

A 'wontfix' is a reasonable answer here, in which case I'll just update the comments.

xavierleroy commented:
Thanks for the measurement.  Searching the Web a bit, I see that Microsoft's crypto API and its implementation changed several times in the past, so my vague recollection of a possible code size problem with a crypto PRNG is certainly outdated, if not plain wrong.

I also see that there's a `rand_s` function in the CRT that draws from the same randomness source as the crypto API, but doesn't require to link with bcrypt.  Three calls to `rand_s` would provide 96 bits of entropy, which is plenty.  What do you think about this?

MisterDA commented:
> I also see that there's a `rand_s` function in the CRT

Not sure why or how I missed this one. I've changed my patch to use [`rand_s`](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/rand-s?view=msvc-170).

nojb reviewed (approved):
LGTM

xavierleroy commented on runtime/win32.c:
```diff
@@ -620,22 +622,33 @@ void caml_win32_unregister_overflow_detection(void)
+      nread += sizeof(unsigned int);
+
+  while (nread > 0) data[n++] = buffer[--nread];
+  /* If the kernel provided enough entropy, we now have 96 bits
```
Nitpicking: it's OK to spread the 3 4-byte random integers over the first 12 integers of `data`, one byte at a time.  But you could just as well write the return values of `rand_s` into `data[0]`, `data[1]` and `data[2]`, and stop there if the 3 calls to `rand_s` succeeded.  In both cases you have 96 bits of entropy, and the Random module will make good use of these bits, no matter how they are spread over the elements of the `data` array.

MisterDA replied:
I've implemented your suggestion, I hope correctly.

xavierleroy reviewed (approved):
You didn't have to change the code, but thanks for having changed it, the result is short and sweet.

I ran a chi^2 test on the output of `caml_sys_random_seed`.  It passes with the new code using `rand_s` , and fails on the old code, which is as expected, and a confirmation that the new code is good.

Now merging...
